### PR TITLE
Improve resolving subject claim when user resides in ancestor organization.

### PR DESCRIPTION
**Pull request by sadilchamishka**
### Proposed changes in this pull request

The subject claim for the given user ID can only be resolved if user exist in that tenant(organization). Therefore the logic is improved to first check the user's existence. If user not found, user should be resolved from the ancestor organizations before trying to resolve the subject claim.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 1]
```diff
@@ -782,9 +786,36 @@ private String getSubjectClaimFromUserStore(String subjectClaimUri, Authenticate
...
+                    .getUserClaimValueWithID(authenticatedUser.getUserId(), subjectClaimUri, null);
+        }
 
-        return userStoreManager
-                .getUserClaimValueWithID(authenticatedUser.getUserId(), subjectClaimUri, null);
+        /*
+            If the user is not in the user store, check if the user can be resolved from ancestor organizations and
+            resolve the subject claim.
+        */
+        try {
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:798]: invoke this only if organization management enabled

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 2]
```diff
@@ -782,9 +786,36 @@ private String getSubjectClaimFromUserStore(String subjectClaimUri, Authenticate
...
+            String organizationId = OAuth2ServiceComponentHolder.getInstance().getOrganizationManager()
+                    .resolveOrganizationId(authenticatedUser.getTenantDomain());
+            Optional<String> userResideOrgId = OAuth2ServiceComponentHolder.getOrganizationUserResidentResolverService()
+                    .resolveResidentOrganization(authenticatedUser.getUserId(), organizationId);
+            if (!userResideOrgId.isPresent()) {
+                throw new org.wso2.carbon.user.core.UserStoreException("Could not resolve the user reside " +
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:804]:
```suggestion
                throw new org.wso2.carbon.user.core.UserStoreException("Could not resolve the user resident " +
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java (hunk 1 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:798]: let's extract this to a private method too

Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java]: Fixed by - https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2103/commits/b3a1b2a1838130a38dc197f0ab9658d22237ab37

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java (hunk 2 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java]: This part were removed by this commit - https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2103/commits/b3a1b2a1838130a38dc197f0ab9658d22237ab37

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 3]
```diff
@@ -1097,4 +1103,35 @@ private static boolean isNotActiveState(String appState) {
...
         return !APP_STATE_ACTIVE.equalsIgnoreCase(appState);
     }
+
+    /**
+     * If the user is not found in the given tenant domain, check the user existence from ancestor organizations and
+     * provide the correct user store manager.
+     *
+     * @param tenantDomain The tenant domain of the authenticated user.
+     * @param userId The ID of the authenticated user.
+     * @return User store manager of the user reside organization.
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:1113]:
```suggestion
     * @return User store manager of the user resident organization.
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 4]
```diff
@@ -1097,4 +1103,35 @@ private static boolean isNotActiveState(String appState) {
...
+     */
+    private Optional<AbstractUserStoreManager> getUserStoreManagerFromUserResideOrganization(String tenantDomain,
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:1115]: What if the user is in a secondary userstore? This method will be loading the primary userstore manager right?
Is that other user store managers will be recursively searched?

Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:1115]: The primary user store manager has the ability to check the user existence in secondary user store chain.
The following set up was tested.
1 - Mark username as the subject identifier in the console.
2 - Plug a secondary user store and assign admin permission for a user.
3 -  Login to the console with that user, and create a suborganization.
4 - Switch to that suborganization.

The ID token was as follows. The username has been resolved properly and could switch to the suborganization successfully.

Screenshot 2023-06-30 at 07 20 55

Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java]: Fixed by - https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2103/commits/1a9a3b2544c31dde5838f01e845435c8937503c8

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**